torch/tensor.py

```diff
@@ -807,31 +807,46 @@ def align_to(self, *names):
             [name for name in names if not is_ellipsis(name)],
             ellipsis_idx)
 
-    def unflatten(self, dim, namedshape):
-        r"""Unflattens the named dimension :attr:`dim`, viewing it in the shape
-        specified by :attr:`namedshape`.
+    def unflatten(self, dim, sizes):
+        r"""Expands the dimension :attr:`dim` of the :attr:`self` tensor over multiple dimensions
+        of sizes given by :attr:`sizes`.
+
+        * :attr:`sizes` is the new shape of the unflattened dimension and it can be a `Iterable[int]` as well
+          as `torch.Size` if :attr:`self` is a `Tensor`, or `namedshape` (iterable of ``(name, size)`` tuples) 
+          if :attr:`self` is a `NamedTensor`. The total number of elements in sizes must match the number 
+          of elements in the original dim being unflattened.
 
         Arguments:
-            namedshape: (iterable of ``(name, size)`` tuples).
+            dim (Union[int, str]): Dimension to unflatten
+            sizes (Union[torch.Size, Iterable[Tuple[str, int]]]): New shape of the unflattened dimension
 
-        Examples::
+        Examples:
+            >>> torch.randn(3, 4, 1).unflatten(1, (2, 2)).shape
+            torch.Size([3, 2, 2, 1])
+            >>> torch.randn(2, 4, names=('A', 'B')).unflatten('B', (('B1', 2), ('B2', 2)))
+            tensor([[[-1.1772,  0.0180],
+                    [ 0.2412,  0.1431]],
 
-            >>> flat_imgs = torch.rand(32, 3 * 128 * 128, names=('N', 'features'))
-            >>> imgs = flat_imgs.unflatten('features', (('C', 3), ('H', 128), ('W', 128)))
-            >>> imgs.names, imgs.shape
-            (('N', 'C', 'H', 'W'), torch.Size([32, 3, 128, 128]))
+                    [[-1.1819, -0.8899],
+                    [ 1.5813,  0.2274]]], names=('A', 'B1', 'B2'))
 
         .. warning::
             The named tensor API is experimental and subject to change.
-
         """
         relevant_args = (self,)
         from torch.overrides import has_torch_function, handle_torch_function
         if type(self) is not Tensor and has_torch_function(relevant_args):
             return handle_torch_function(Tensor.unflatten, relevant_args, self, dim, namedshape)
-        names, sizes = unzip_namedshape(namedshape)
+
+        names = None
+        if isinstance(dim, str) or self.has_names():
+            try:
+                names, sizes = unzip_namedshape(sizes)
+            except TypeError:
+                raise TypeError("unflatten: Expected sizes to be of type Iterable[Tuple[str, int]] but got " + type(sizes).__name__)
         return super(Tensor, self).unflatten(dim, sizes, names)
```

Comment:
Sorry, I wasn't clear here with my previous comment.

Before this PR, we supported the following:
- unflatten(Tensor, int, namedshape)
- unflatten(NamedTensor, int, namedshape)
- unflatten(Tensor, str, namedshape)
- unflatten(NamedTensor, str, namedshape)

It looks like we've dropped support for the case of
- unflatten(Tensor, int, namedshape)

Ideally, we'd keep BC. My suggestion here is the following:
- check if `sizes` is a namedshape.
    - if it is a namedshape, then parse it into `sizes` and `names`
    - if it's not a namedshape, then pass in `sizes` and `names = None`
- if `sizes` is not a namedshape and the input is a named tensor, then we raise an error. Previously we had required `sizes` to always be a namedshape. Now we are relaxing that constraint. Since we are adding new behavior here, this error isn't BC-breaking

Reply:
NB: It's not the biggest problem in the world if we break BC here. I don't think many users were using the `unflatten(Tensor, int, namedshape)` case, and we advertise named tensors as "experimental and subject to change without warning".